google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java

```diff
@@ -1551,6 +1588,193 @@ public void testSessionMetrics() throws Exception {
+      verifyMetricData(metricDataCollection, NUM_ACQUIRED_SESSIONS_NAME, 1, 2L);
+
+      // Max in use session are 2.
+      verifyMetricData(metricDataCollection, MAX_IN_USE_SESSIONS_NAME, 1, 2D);
+
+      // Max Allowed sessions should be 3
+      verifyMetricData(metricDataCollection, MAX_ALLOWED_SESSIONS_NAME, 1, 3D);
+
+      // Released sessions should be 0
+      verifyMetricData(metricDataCollection, NUM_RELEASED_SESSIONS_NAME, 1, 0L);
+
+      // Num sessions in pool
+      verifyMetricData(
+          metricDataCollection, NUM_SESSIONS_IN_POOL_NAME, 1, NUM_SESSIONS_IN_USE_NAME, 2);
+
+      PooledSessionFuture session3 = pool.getSession();
+      session3.get();
+
+      final CountDownLatch latch = new CountDownLatch(1);
+      // Try asynchronously to take another session. This attempt should time out.
+      Future<Void> fut =
+          executor.submit(
+              () -> {
+                latch.countDown();
+                Session session = pool.getSession();
+                session.close();
+                return null;
+              });
+      // Wait until the background thread is actually waiting for a session.
+      latch.await();
+      // Wait until the request has timed out.
+      int waitCount = 0;
+      while (pool.getNumWaiterTimeouts() == 0L && waitCount < 1000) {
+        Thread.sleep(5L);
+        waitCount++;
+      }
+      // Return the checked out session to the pool so the async request will get a session and
+      // finish.
+      session2.close();
+      // Verify that the async request also succeeds.
+      fut.get(10L, TimeUnit.SECONDS);
+      executor.shutdown();
+
+      metricDataCollection = inMemoryMetricReader.collectAllMetrics();
+
+      // Max Allowed sessions should be 3
+      verifyMetricData(metricDataCollection, MAX_ALLOWED_SESSIONS_NAME, 1, 3D);
+
+      // Session timeouts 1
+      verifyMetricData(metricDataCollection, GET_SESSION_TIMEOUTS_NAME, 1, 1L);
+
+      // Max in use session are 2.
+      verifyMetricData(metricDataCollection, MAX_IN_USE_SESSIONS_NAME, 1, 3D);
+
+      // Session released 2
+      verifyMetricData(metricDataCollection, NUM_RELEASED_SESSIONS_NAME, 1, 2L);
+
+      // Acquired sessions are 4.
+      verifyMetricData(metricDataCollection, NUM_ACQUIRED_SESSIONS_NAME, 1, 4L);
+
+      // Num sessions in pool
+      verifyMetricData(
+          metricDataCollection, NUM_SESSIONS_IN_POOL_NAME, 1, NUM_SESSIONS_IN_USE_NAME, 2);
+      verifyMetricData(
+          metricDataCollection, NUM_SESSIONS_IN_POOL_NAME, 1, NUM_SESSIONS_AVAILABLE_NAME, 1);
+    }
+  }
+
+  private static void verifyMetricData(
+      Collection<MetricData> metricDataCollection, String metricName, int size, long value) {
+    Collection<MetricData> metricDataFiltered =
+        metricDataCollection.stream()
+            .filter(x -> x.getName().equals(metricName))
+            .collect(Collectors.toList());
+    ;
+    assertThat(metricDataFiltered.size()).isEqualTo(size);
```

Comment:
nit: we prefer the use of `assertEquals(..)` over `assertThat(..).isEqualTo(..)`. This test class is relatively old, so I know that there are a lot of `assertThat(...)` calls here, but we are trying to remove those step-by-step.